extensions/amp-story/1.0/progress-bar.js

```diff
@@ -119,10 +177,180 @@ export class ProgressBar {
+      .querySelector('amp-story-page')
+      .getBoundingClientRect()./*OK*/ width;
+
+    this.vsync_.mutate(() => {
+      // Don't animate the progress bar in pageload.
+      this.timer_.delay(() => {
+        this.root_.classList.add('i-amphtml-animate-progress');
+      }, 60);
+    });
+
     this.isBuilt_ = true;
     return this.getRoot();
   }
 
+  /**
+   * Reacts to story replay.
+   * @private
+   */
+  replay_() {
+    this.progressBarIndex_ = 0;
+    this.render_();
+  }
+
+  /**
+   * Renders the segments by setting their corresponding width and translate.
+   * @private
+   */
+  render_() {
+    const segmentWidth = this.getSegmentWidth_();
+    let translateX =
+      -(this.progressBarIndex_ - this.getPrevEllipsisCount_()) *
+      (ELLIPSE_WIDTH_PX + SEGMENTS_MARGIN_PX);
+
+    for (let index = 0; index < this.segmentCount_; index++) {
+      const width =
+        index >= this.progressBarIndex_ &&
+        index < this.progressBarIndex_ + MAX_SEGMENTS
+          ? segmentWidth
+          : ELLIPSE_WIDTH_PX;
+      this.transform_(this.segments_[index], translateX, width);
+      translateX += width + SEGMENTS_MARGIN_PX;
+    }
+  }
+
+  /**
+   * Applies transform to a segment.
+   * @param {!Element} segment
+   * @param {number} translateX
+   * @param {number} width
+   * @private
+   */
+  transform_(segment, translateX, width) {
+    if (this.storeService_.get(StateProperty.RTL_STATE)) {
+      translateX *= -1;
+    }
+    this.vsync_.mutate(() => {
+      // Do not remove translateZ(0.00001px) as it prevents an iOS repaint issue.
+      // http://mir.aculo.us/2011/12/07/the-case-of-the-disappearing-element/
+      segment.setAttribute(
+        'style',
+        `transform: translate3d(${translateX}px, 0px, 0.00001px) scaleX(${width /
+          ELLIPSE_WIDTH_PX});`
+      );
+    });
+  }
+
+  /**
+   * Gets the individual segment width.
+   * @return {number}
+   * @private
+   */
+  getSegmentWidth_() {
+    const nextEllipsisCount = this.getNextEllipsisCount_();
+    const prevEllipsisCount = this.getPrevEllipsisCount_();
+    const totalEllipsisWidth =
+      (nextEllipsisCount + prevEllipsisCount) *
+      (ELLIPSE_WIDTH_PX + SEGMENTS_MARGIN_PX);
+    const totalSegmentsWidth = this.barWidthPx_ - totalEllipsisWidth;
+
+    return (
+      totalSegmentsWidth / Math.min(this.segmentCount_, MAX_SEGMENTS) -
+      SEGMENTS_MARGIN_PX
+    );
+  }
+
+  /**
+   * Gets the count of ellipsis to the right (in LTR).
+   * @return {number}
+   * @private
+   */
+  getNextEllipsisCount_() {
+    const nextPagesCount =
+      this.segmentCount_ - (this.progressBarIndex_ + MAX_SEGMENTS);
+    return nextPagesCount > 3 ? 3 : Math.max(nextPagesCount, 0);
+  }
+
+  /**
+   * Gets the count of ellipsis to the left (in LTR).
+   * @return {number}
+   * @private
+   */
+  getPrevEllipsisCount_() {
+    return this.progressBarIndex_ > 3
+      ? 3
+      : Math.min(this.increment_, this.progressBarIndex_);
```

Comment:
I think `return Math.min(3, this.progressBarIndex_);` would solve this?

This way, you could remove the `this.increment_` ivar and keep it as a local variable in `checkIndexForOverflow_`.

Reply:
Done.